src/main/java/org/opensearch/security/auth/UserInjector.java

```diff
@@ -188,15 +167,15 @@ boolean injectUser(SecurityRequestChannel request) {
             threadPool.getThreadContext().putTransient(ConfigConstants.OPENDISTRO_SECURITY_REMOTE_ADDRESS, xffResolver.resolve(request));
         }
 
-        threadPool.getThreadContext().putTransient(ConfigConstants.OPENDISTRO_SECURITY_USER, injectedUser);
-        auditLog.logSucceededLogin(injectedUser.getName(), true, null, request);
+        threadPool.getThreadContext().putTransient(ConfigConstants.OPENDISTRO_SECURITY_USER, injectedUser.getUser());
+        auditLog.logSucceededLogin(injectedUser.getUser().getName(), true, null, request);
 
         return true;
     }
 
-    protected Map<String, String> mapFromArray(String... keyValues) {
+    protected ImmutableMap<String, String> mapFromArray(String... keyValues) {
         if (keyValues == null) {
-            return Map.of();
+            return ImmutableMap.of();
```

Comment:
FYI Since the min java compatibility version is now JDK21, we can take advantage of immutable classes from within the JDK: https://docs.oracle.com/en/java/javase/21/core/creating-immutable-lists-sets-and-maps.html

`Map.of()` now creates an immutable map and gives an UnsupportedOperationException if you try to put a new entry into the map.

Reply:
The only downside I see of using JDK APIs is that there is no way for the consumer to know if a map is immutable or not until actually trying to make a modification at which point they would get a RuntimeException.

Reply:
These are also my thoughts, using the explicit ImmutableSet class, I document that something's is immutable and I can also trust that something's immutable if it is passed as argument. 

That saves quite a bit of copying/wrapping because one is afraid that the object might be mutated somewhere else.